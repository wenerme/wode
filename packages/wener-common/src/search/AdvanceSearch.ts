import { formatAdvanceSearch } from './formatAdvanceSearch';
import { optimizeAdvanceSearch } from './optimizeAdvanceSearch';
import { parseAdvanceSearch } from './parseAdvanceSearch';
import type * as types from './types';

export namespace AdvanceSearch {
  export type Exprs = types.Exprs;
  export type Expr = types.Expr;
  export type Value = types.Value;

  export const parse = parseAdvanceSearch;
  export const format = formatAdvanceSearch;
  export const optimize = optimizeAdvanceSearch;
}
