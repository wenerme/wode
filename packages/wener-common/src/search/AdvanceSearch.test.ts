import { describe, expect, it } from 'vitest';
import { AdvanceSearch } from './AdvanceSearch';

describe('AdvanceSearch', () => {
  it('should parse', () => {
    for (const input of [
      //
      '-a',
      'a',
      'a b',
      'a -b',
      'a-b',
      'a"b',
      'a&b',
      'NOT a',
      'NOT -a',
      'HELLO -WORLD',
      '(a)',
      '( a OR B )',
      'A OR B',
      'a:[1,2]',
      'a:*..1',
      'a:ok a:=1 a:>1 a:<1 a:>=1 a:<=1 a:!=1 a:1..2 a:*..1 a:1..* a:[1,2] a:(1,2) a:[1,2) a:(1,2]',
      'NOT (A B) AND (a:ok AND size:>1)',
      'NOT -a',
      'NOT (NOT -a)',
      'owner:@me owner:=@me owner:!=@me',
      '@AI:"Where is my car"',
      '/**/ a',
      '/* Hint */ a',
      '/* a */ a /* b */',
    ]) {
      let out = AdvanceSearch.parse(input);
      let formated = AdvanceSearch.format(out);
      expect(formated, `reformat`).toMatchSnapshot();
      let optimized = AdvanceSearch.format(AdvanceSearch.optimize(out));
      expect(optimized, `optimized`).toMatchSnapshot();
      expect(out, `parsed`).toMatchSnapshot();
      expect(AdvanceSearch.parse(formated), `reformat match original`).toMatchObject(out);
    }
  });
  it('should parse as expected', () => {
    type Case = [string | null | undefined, AdvanceSearch.Expr[]];

    const cases: Case[] = [
      [null, []],
      [undefined, []],
      ['', []],
      // fast path
      ['a', [{ type: 'keyword', value: 'a' }]],
      ['a-b', [{ type: 'keyword', value: 'a-b' }]],
      ['a"b', [{ type: 'keyword', value: 'a"b' }]],
      [
        'hello world',
        [
          { type: 'keyword', value: 'hello' },
          { type: 'keyword', value: 'world' },
        ],
      ],
      // advance
      ['-a', [{ type: 'keyword', value: 'a', negative: true }]],
      [
        '/*Hi*/ hello -world',
        [
          { type: 'comment', value: 'Hi' },
          { type: 'keyword', value: 'hello', negative: false },
          {
            type: 'keyword',
            value: 'world',
            negative: true,
          },
        ],
      ],
      ['"Hello"', [{ type: 'keyword', value: 'Hello', exact: true }]],
      ['-"Hello"', [{ type: 'keyword', value: 'Hello', exact: true, negative: true }]],
      ['is:ok', [{ type: 'compare', field: 'is', operator: 'match', value: { value: 'ok' } }]],
      ['-is:ok', [{ type: 'compare', field: 'is', negative: true, operator: 'match', value: { value: 'ok' } }]],
    ];

    for (const [input, expected] of cases) {
      let out = AdvanceSearch.parse(input);
      expect(out).toMatchObject(expected);
      expect(AdvanceSearch.parse(AdvanceSearch.format(out)), 'reformat should match').toMatchObject(expected);
    }
  });

  it('should optimize simple', () => {
    type Case = {
      input: AdvanceSearch.Expr[];
      expected: AdvanceSearch.Expr[];
    };
    const cases: Case[] = [
      // rm empty comment
      {
        input: [
          { type: 'comment', value: '' },
          { type: 'keyword', value: 'a' },
        ],
        expected: [{ type: 'keyword', value: 'a' }],
      },
      // unwrap parentheses
      {
        input: [
          { type: 'parentheses', value: [{ type: 'keyword', value: 'a' }] },
          { type: 'keyword', value: 'b' },
        ],
        expected: [
          { type: 'keyword', value: 'a' },
          { type: 'keyword', value: 'b' },
        ],
      },
      // not not
      {
        input: [{ type: 'not', value: { type: 'not', value: { type: 'keyword', value: 'a' } } }],
        expected: [{ type: 'keyword', value: 'a', negative: false }],
      },
      // not to negative
      {
        input: [{ type: 'not', value: { type: 'keyword', value: 'a' } }],
        expected: [{ type: 'keyword', value: 'a', negative: true }],
      },
    ];

    for (let i = 0; i < cases.length; i++) {
      const { input, expected } = cases[i];
      let out = AdvanceSearch.optimize(input);
      expect(out, `case #${i}`).toEqual(expected);
    }
  });

  it('should optimize by formated', () => {
    type Case = [string, string];

    const cases: Case[] = [
      ['( a )', 'a'],
      ['NOT a', '-a'],
      ['NOT -a', 'a'],
      ['NOT is:ok', '-is:ok'],
      ['NOT -is:ok', 'is:ok'],
      ['NOT -is:=ok', 'is:=ok'],
      ['NOT is:=ok', 'is:!=ok'],
      ['NOT (NOT -a)', '-a'],
    ];

    for (const [input, expected] of cases) {
      let out = AdvanceSearch.optimize(AdvanceSearch.parse(input));
      expect(AdvanceSearch.format(out), `${input} -> ${expected}: ${JSON.stringify(out)}`).toEqual(expected);
    }
  });

  it.fails('should parse parentheses', () => {
    let out = AdvanceSearch.parse('(a)');
    console.log(out);
    expect(out).toEqual([{ type: 'parentheses', value: [{ type: 'keyword', value: 'a' }] }]);
  });
});
