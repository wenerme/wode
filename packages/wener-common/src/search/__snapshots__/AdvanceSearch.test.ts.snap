// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AdvanceSearch > should parse 1`] = `"-a"`;

exports[`AdvanceSearch > should parse 2`] = `"-a"`;

exports[`AdvanceSearch > should parse 3`] = `
[
  {
    "negative": true,
    "type": "keyword",
    "value": "a",
  },
]
`;

exports[`AdvanceSearch > should parse 4`] = `"a"`;

exports[`AdvanceSearch > should parse 5`] = `"a"`;

exports[`AdvanceSearch > should parse 6`] = `
[
  {
    "type": "keyword",
    "value": "a",
  },
]
`;

exports[`AdvanceSearch > should parse 7`] = `"a b"`;

exports[`AdvanceSearch > should parse 8`] = `"a b"`;

exports[`AdvanceSearch > should parse 9`] = `
[
  {
    "type": "keyword",
    "value": "a",
  },
  {
    "type": "keyword",
    "value": "b",
  },
]
`;

exports[`AdvanceSearch > should parse 10`] = `"a -b"`;

exports[`AdvanceSearch > should parse 11`] = `"a -b"`;

exports[`AdvanceSearch > should parse 12`] = `
[
  {
    "negative": false,
    "type": "keyword",
    "value": "a",
  },
  {
    "negative": true,
    "type": "keyword",
    "value": "b",
  },
]
`;

exports[`AdvanceSearch > should parse 13`] = `"a-b"`;

exports[`AdvanceSearch > should parse 14`] = `"a-b"`;

exports[`AdvanceSearch > should parse 15`] = `
[
  {
    "negative": false,
    "type": "keyword",
    "value": "a-b",
  },
]
`;

exports[`AdvanceSearch > should parse 16`] = `"a"b"`;

exports[`AdvanceSearch > should parse 17`] = `"a"b"`;

exports[`AdvanceSearch > should parse 18`] = `
[
  {
    "negative": false,
    "type": "keyword",
    "value": "a"b",
  },
]
`;

exports[`AdvanceSearch > should parse 19`] = `"a&b"`;

exports[`AdvanceSearch > should parse 20`] = `"a&b"`;

exports[`AdvanceSearch > should parse 21`] = `
[
  {
    "type": "keyword",
    "value": "a&b",
  },
]
`;

exports[`AdvanceSearch > should parse 22`] = `"NOT a"`;

exports[`AdvanceSearch > should parse 23`] = `"-a"`;

exports[`AdvanceSearch > should parse 24`] = `
[
  {
    "type": "not",
    "value": {
      "negative": false,
      "type": "keyword",
      "value": "a",
    },
  },
]
`;

exports[`AdvanceSearch > should parse 25`] = `"NOT -a"`;

exports[`AdvanceSearch > should parse 26`] = `"a"`;

exports[`AdvanceSearch > should parse 27`] = `
[
  {
    "type": "not",
    "value": {
      "negative": true,
      "type": "keyword",
      "value": "a",
    },
  },
]
`;

exports[`AdvanceSearch > should parse 28`] = `"HELLO -WORLD"`;

exports[`AdvanceSearch > should parse 29`] = `"HELLO -WORLD"`;

exports[`AdvanceSearch > should parse 30`] = `
[
  {
    "negative": false,
    "type": "keyword",
    "value": "HELLO",
  },
  {
    "negative": true,
    "type": "keyword",
    "value": "WORLD",
  },
]
`;

exports[`AdvanceSearch > should parse 31`] = `"(a)"`;

exports[`AdvanceSearch > should parse 32`] = `"a"`;

exports[`AdvanceSearch > should parse 33`] = `
[
  {
    "type": "parentheses",
    "value": [
      {
        "negative": false,
        "type": "keyword",
        "value": "a",
      },
    ],
  },
]
`;

exports[`AdvanceSearch > should parse 34`] = `"(a OR B)"`;

exports[`AdvanceSearch > should parse 35`] = `"a OR B"`;

exports[`AdvanceSearch > should parse 36`] = `
[
  {
    "type": "parentheses",
    "value": [
      {
        "operator": "or",
        "type": "logical",
        "value": [
          {
            "negative": false,
            "type": "keyword",
            "value": "a",
          },
          {
            "negative": false,
            "type": "keyword",
            "value": "B",
          },
        ],
      },
    ],
  },
]
`;

exports[`AdvanceSearch > should parse 37`] = `"A OR B"`;

exports[`AdvanceSearch > should parse 38`] = `"A OR B"`;

exports[`AdvanceSearch > should parse 39`] = `
[
  {
    "operator": "or",
    "type": "logical",
    "value": [
      {
        "negative": false,
        "type": "keyword",
        "value": "A",
      },
      {
        "negative": false,
        "type": "keyword",
        "value": "B",
      },
    ],
  },
]
`;

exports[`AdvanceSearch > should parse 40`] = `"a:[1,2]"`;

exports[`AdvanceSearch > should parse 41`] = `"a:[1,2]"`;

exports[`AdvanceSearch > should parse 42`] = `
[
  {
    "field": "a",
    "operator": "range",
    "type": "compare",
    "value": {
      "maximum": {
        "type": "literal",
        "value": 2,
      },
      "maximumExclusive": false,
      "minimum": {
        "type": "literal",
        "value": 1,
      },
      "minimumExclusive": false,
      "type": "range",
    },
  },
]
`;

exports[`AdvanceSearch > should parse 43`] = `"a:*..1"`;

exports[`AdvanceSearch > should parse 44`] = `"a:*..1"`;

exports[`AdvanceSearch > should parse 45`] = `
[
  {
    "field": "a",
    "operator": "range",
    "type": "compare",
    "value": {
      "maximum": {
        "type": "literal",
        "value": 1,
      },
      "minimum": undefined,
      "type": "range",
    },
  },
]
`;

exports[`AdvanceSearch > should parse 46`] = `"a:ok a:=1 a:>1 a:<1 a:>=1 a:<=1 a:!=1 a:1..2 a:*..1 a:1..* a:[1,2] a:(1,2) a:[1,2) a:(1,2]"`;

exports[`AdvanceSearch > should parse 47`] = `"a:ok a:=1 a:>1 a:<1 a:>=1 a:<=1 a:!=1 a:1..2 a:*..1 a:1..* a:[1,2] a:(1,2) a:[1,2) a:(1,2]"`;

exports[`AdvanceSearch > should parse 48`] = `
[
  {
    "field": "a",
    "operator": "match",
    "type": "compare",
    "value": {
      "type": "literal",
      "value": "ok",
    },
  },
  {
    "field": "a",
    "operator": "eq",
    "type": "compare",
    "value": {
      "type": "literal",
      "value": 1,
    },
  },
  {
    "field": "a",
    "operator": "gt",
    "type": "compare",
    "value": {
      "type": "literal",
      "value": 1,
    },
  },
  {
    "field": "a",
    "operator": "lt",
    "type": "compare",
    "value": {
      "type": "literal",
      "value": 1,
    },
  },
  {
    "negative": false,
    "type": "keyword",
    "value": "a:>=1",
  },
  {
    "negative": false,
    "type": "keyword",
    "value": "a:<=1",
  },
  {
    "field": "a",
    "operator": "ne",
    "type": "compare",
    "value": {
      "type": "literal",
      "value": 1,
    },
  },
  {
    "field": "a",
    "operator": "range",
    "type": "compare",
    "value": {
      "maximum": {
        "type": "literal",
        "value": 2,
      },
      "minimum": {
        "type": "literal",
        "value": 1,
      },
      "type": "range",
    },
  },
  {
    "field": "a",
    "operator": "range",
    "type": "compare",
    "value": {
      "maximum": {
        "type": "literal",
        "value": 1,
      },
      "minimum": undefined,
      "type": "range",
    },
  },
  {
    "field": "a",
    "operator": "range",
    "type": "compare",
    "value": {
      "maximum": undefined,
      "minimum": {
        "type": "literal",
        "value": 1,
      },
      "type": "range",
    },
  },
  {
    "field": "a",
    "operator": "range",
    "type": "compare",
    "value": {
      "maximum": {
        "type": "literal",
        "value": 2,
      },
      "maximumExclusive": false,
      "minimum": {
        "type": "literal",
        "value": 1,
      },
      "minimumExclusive": false,
      "type": "range",
    },
  },
  {
    "field": "a",
    "operator": "range",
    "type": "compare",
    "value": {
      "maximum": {
        "type": "literal",
        "value": 2,
      },
      "maximumExclusive": true,
      "minimum": {
        "type": "literal",
        "value": 1,
      },
      "minimumExclusive": true,
      "type": "range",
    },
  },
  {
    "field": "a",
    "operator": "range",
    "type": "compare",
    "value": {
      "maximum": {
        "type": "literal",
        "value": 2,
      },
      "maximumExclusive": true,
      "minimum": {
        "type": "literal",
        "value": 1,
      },
      "minimumExclusive": false,
      "type": "range",
    },
  },
  {
    "field": "a",
    "operator": "range",
    "type": "compare",
    "value": {
      "maximum": {
        "type": "literal",
        "value": 2,
      },
      "maximumExclusive": false,
      "minimum": {
        "type": "literal",
        "value": 1,
      },
      "minimumExclusive": true,
      "type": "range",
    },
  },
]
`;

exports[`AdvanceSearch > should parse 49`] = `"NOT (A B) AND (a:ok AND size:>1)"`;

exports[`AdvanceSearch > should parse 50`] = `"NOT (A B) AND (a:ok AND size:>1)"`;

exports[`AdvanceSearch > should parse 51`] = `
[
  {
    "operator": "and",
    "type": "logical",
    "value": [
      {
        "type": "not",
        "value": {
          "type": "parentheses",
          "value": [
            {
              "negative": false,
              "type": "keyword",
              "value": "A",
            },
            {
              "negative": false,
              "type": "keyword",
              "value": "B",
            },
          ],
        },
      },
      {
        "type": "parentheses",
        "value": [
          {
            "operator": "and",
            "type": "logical",
            "value": [
              {
                "field": "a",
                "operator": "match",
                "type": "compare",
                "value": {
                  "type": "literal",
                  "value": "ok",
                },
              },
              {
                "field": "size",
                "operator": "gt",
                "type": "compare",
                "value": {
                  "type": "literal",
                  "value": 1,
                },
              },
            ],
          },
        ],
      },
    ],
  },
]
`;

exports[`AdvanceSearch > should parse 52`] = `"NOT -a"`;

exports[`AdvanceSearch > should parse 53`] = `"a"`;

exports[`AdvanceSearch > should parse 54`] = `
[
  {
    "type": "not",
    "value": {
      "negative": true,
      "type": "keyword",
      "value": "a",
    },
  },
]
`;

exports[`AdvanceSearch > should parse 55`] = `"NOT (NOT -a)"`;

exports[`AdvanceSearch > should parse 56`] = `"-a"`;

exports[`AdvanceSearch > should parse 57`] = `
[
  {
    "type": "not",
    "value": {
      "type": "parentheses",
      "value": [
        {
          "type": "not",
          "value": {
            "negative": true,
            "type": "keyword",
            "value": "a",
          },
        },
      ],
    },
  },
]
`;

exports[`AdvanceSearch > should parse 58`] = `"owner:=@me owner:=@me owner:!=@me"`;

exports[`AdvanceSearch > should parse 59`] = `"owner:=@me owner:=@me owner:!=@me"`;

exports[`AdvanceSearch > should parse 60`] = `
[
  {
    "field": "owner",
    "operator": "eq",
    "type": "compare",
    "value": {
      "format": "mention",
      "value": "me",
    },
  },
  {
    "field": "owner",
    "operator": "eq",
    "type": "compare",
    "value": {
      "format": "mention",
      "value": "me",
    },
  },
  {
    "field": "owner",
    "operator": "ne",
    "type": "compare",
    "value": {
      "format": "mention",
      "value": "me",
    },
  },
]
`;

exports[`AdvanceSearch > should parse 61`] = `"@AI:"Where is my car""`;

exports[`AdvanceSearch > should parse 62`] = `"@AI:"Where is my car""`;

exports[`AdvanceSearch > should parse 63`] = `
[
  {
    "field": "AI",
    "mention": true,
    "operator": "match",
    "type": "compare",
    "value": {
      "value": "Where is my car",
    },
  },
]
`;

exports[`AdvanceSearch > should parse 64`] = `"/*  */ a"`;

exports[`AdvanceSearch > should parse 65`] = `"a"`;

exports[`AdvanceSearch > should parse 66`] = `
[
  {
    "type": "comment",
    "value": "",
  },
  {
    "negative": false,
    "type": "keyword",
    "value": "a",
  },
]
`;

exports[`AdvanceSearch > should parse 67`] = `"/* Hint */ a"`;

exports[`AdvanceSearch > should parse 68`] = `"/* Hint */ a"`;

exports[`AdvanceSearch > should parse 69`] = `
[
  {
    "type": "comment",
    "value": "Hint",
  },
  {
    "negative": false,
    "type": "keyword",
    "value": "a",
  },
]
`;

exports[`AdvanceSearch > should parse 70`] = `"/* a */ a /* b */"`;

exports[`AdvanceSearch > should parse 71`] = `"/* a */ a /* b */"`;

exports[`AdvanceSearch > should parse 72`] = `
[
  {
    "type": "comment",
    "value": "a",
  },
  {
    "negative": false,
    "type": "keyword",
    "value": "a",
  },
  {
    "type": "comment",
    "value": "b",
  },
]
`;
